### DISCUSSION TITLE
Routes?

### DISCUSSION THREAD

**Discussion by oyvindsolstad** [javalent/obsidian-leaflet#101] [Ideas]
(sorry for starting so many threads)

I'm looking at the way Wikipedia show routes on maps. As I work with travel, it's super interesting to look at ways to visualize data inside Obsidian. 

Here's the map Wikipedia makes for I-80 (Interstate 80): https://en.wikipedia.org/wiki/Interstate_80#/map/0

![Skjermbilde 2021-06-13 kl  18 14 42](https://user-images.githubusercontent.com/4631971/121814816-47a17100-cc73-11eb-9012-d3010f7636bf.png)

The map is zoomable too:

![Skjermbilde 2021-06-13 kl  18 14 57](https://user-images.githubusercontent.com/4631971/121814822-51c36f80-cc73-11eb-8a7a-b8fd3fac41c0.png)

Is there any way to make the leaflet plugin do similar? I understand that a route needs lots of datapoints to be zoomable. 

It would be useful for all kinds of stuff I do: 
• map routes 
• plan travels 
• show past travels
• production planning 

etc.

**Comment by valentine195** [comment]
This could possibly be done with GeoJSON Files. See this issue: https://github.com/valentine195/obsidian-leaflet-plugin/issues/93

Edit: to be clear, I haven't implemented this.

**Reply by oyvindsolstad** [reply]
Looks like a way to do it. I guess as a developer it all boils down to "make features most will use" vs "cool advanced stuff four people will love". 😄

GeoJSON looks interesting, cool format. Reading up [here](https://leafletjs.com/examples/geojson/).

**Reply by valentine195** [reply]
I added the `geojson` tag. Check it out in the Readme!

**Reply by oyvindsolstad** [reply]
Fantastic! One small question: How do I get the json/geojson-files into Obsidian? Do I have to link to external files?

**Reply by valentine195** [reply]
Sorry - it needs to be in your vault, then you can wiki link to it.

**Reply by oyvindsolstad** [reply]
That works flawlessly! Wow. I downloaded [a neighbourhood map file for Toronto](https://handsondataviz.org/geojsonio.html) just to test. I'm so impressed by your coding skills and plugin. Thank you again. I'm going to use this so much. 

![Skjermbilde 2021-06-15 kl  13 38 41](https://user-images.githubusercontent.com/4631971/122046427-2d84a180-cddf-11eb-8d56-536d673f717f.png)

**Comment by oyvindsolstad** [comment]
It worked! I looked at the wiki posts for the highways and found KML files. They were not to easy to download in a useful format so I just grabbed all the points, used an MacOS app called [Clean Text](https://www.apimac.com/mac/cleantext/), made a set of macros that changes the formats to be the way JSON wants it to be. Saved it in the "filer" (files in Norwegian) and added it to the highway map. 

Voila. 

![Skjermbilde 2021-06-15 kl  17 12 53](https://user-images.githubusercontent.com/4631971/122078593-ff15bf00-cdfc-11eb-9c0a-f6e902a0fcbc.png)

All zoomable and lovely. 


![Skjermbilde 2021-06-15 kl  17 14 04](https://user-images.githubusercontent.com/4631971/122078795-22406e80-cdfd-11eb-9ff8-af1110275a4c.png)
![Skjermbilde 2021-06-15 kl  17 14 12](https://user-images.githubusercontent.com/4631971/122078802-23719b80-cdfd-11eb-96ca-b201e2ba1d48.png)

**Reply by valentine195** [reply]
Wonderful!

@Moonbase59 showed me this [GeoJSON styling](https://github.com/mapbox/simplestyle-spec/tree/master/1.1.0) specification, too, so I implemented it in version **3.18.0**, if you wanted to style the line.

Edit: Also, I added the `geojsonColor` parameter for a simple color change.

**Reply by oyvindsolstad** [reply]
YEY! I was wondering about that, as i tried and couldn't change the colors. I had that on "you made an error", not on "the plugin doesn't support it" tbh.

**Reply by oyvindsolstad** [reply]
That worked perfectly too. 

If I want to have several routes on the same map, do I have to merge JSON-files?

**Reply by valentine195** [reply]
No, you should be able to specify any number of GeoJSON files under the `geojson` parameter. Please see [this](https://github.com/valentine195/obsidian-leaflet-plugin#geojson) section of the ReadMe.

**Reply by oyvindsolstad** [reply]
Love it. Worked that one too. 

![Skjermbilde 2021-06-15 kl  17 56 02](https://user-images.githubusercontent.com/4631971/122085402-f7f1af80-ce02-11eb-8784-453e6341486e.png)

**Reply by oyvindsolstad** [reply]
I just realised that Foursquare has an export function. And I've been checking in to Swarm for close to 10 years. 

Fun project coming up.

**Reply by zhubib1234** [reply]
Can you share your geojson file? I haven't made it. thank you.

**Reply by Moonbase59** [reply]
@oyvindsolstad: It’s rather easy converting GPX data (from trackers) and KML (not KMX!) data from other mapping applications into GeoJSON, as I hinted at in https://github.com/valentine195/obsidian-leaflet-plugin/issues/121#issue-925121587. ;-) I use that strategy to convert personal routes and trails from Google Maps.

Next step after editing: Use a [GeoJSON Linter](https://geojsonlint.com/) to verify the files are okay.

I actually wonder why Wikipedia/Wikimedia make it so hard getting at their maps. Their map data is usually licensed CC-BY 4.0 anyway. Maybe you’d like to open a feature request over there?

Oh, and be aware it pays to re-compress (using [a few lines of Python](https://github.com/valentine195/obsidian-leaflet-plugin/discussions/108#discussion-3414795), by example) the files again after editing, otherwise, with including a few of them for your interstate map, you’ll end up eating _enormous_ amounts of RAM and Obsidian will get _real slow_!

@zhubib1234: Please stop nagging others about "sharing" and start learning how do things _yourself_. There has been enough help already! Should it only be that you don’t understand English so well, I apologize for being rude. You can use <https://github.com/valentine195/obsidian-leaflet-plugin/files/6671787/geo-usa-route66-edit2.geojson.zip> as a starting point for your experiments.

**Reply by oyvindsolstad** [reply]
@Moonbase59 Thanks! I'll have a look, and see if I figure it out. Been meaning to learn some Python to do stuff like this, as most online tools are both slow, have limits on both size and number of conversions - and often don't work. So I need to learn how to do this manually on my own Mac.

**Reply by oyvindsolstad** [reply]
Regarding why Wikipedia/Wikimedia make it so hard getting at their maps. I talked to a former admin, and he said it's a bit of a mess, the whole map thing. Lots of maps are PNGs or SVGs, some are new, quality accurate data, and some are very rudimentary maps with lots of data missing to make load times quicker. 

Would probably be better if they just let Openstreetmap handle that part, but maybe that's not possible with their platform. 

Anyway, with the solution at Mapbox, I can convert GPX to GeoJSON, and I'm good. So thanks for the link!

**Reply by zhubib1234** [reply]
I used “geo USA Route66 edit2.geojson”, but there was no display on the map, and the “simplestyle spec” example given was used, and it was not shown on the map. So I want to learn from the implemented examples.

**Reply by Moonbase59** [reply]
Ok, @zhubib1234, if you have the `geo-usa-route66-edit2.geojson` somewhere in your Obsidian vault, a map like this in your Obsidian note should show it:

````
# U.S. Route 66

```leaflet
id: route66
maxZoom: 18
zoomDelta: 0.5
zoomFeatures: true
geojson: 
  - [[geo-usa-route66-edit2.geojson]]
```
````

![Auswahl_024](https://user-images.githubusercontent.com/3706922/122646381-20d5b580-d11f-11eb-970a-69a31afd03fe.png)

**Reply by Moonbase59** [reply]
> So I need to learn how to do this manually on my own Mac.

Good plan! I don’t know if MacOS comes with Python3, if not, you could use [Homebrew](https://brew.sh/) to easily install it.

It’s good to use Python for that, because it’s relatively easy to learn, and _fast_, even on large datasets. Plus, you can easily make a GoeJSON file "readable" (by using `indent`), and afterwards compress it again (using _no_ `indent`). A text editor is almost _always_ much too slow for work like that.

For the examples, I didn’t even bother to write a script—just fire up `python3` (it might be called just `python`) and _type in_ the commands. And bingo!—GeoJSON edited! Perfect playground. ;-)

**Reply by zhubib1234** [reply]
thanks！

**Comment by oyvindsolstad** [comment]
I've used Google Maps, and [this service](https://mapstogpx.com) to convert links to GPX. Then I use the [toGeoJSON](https://mapbox.github.io/togeojson/) tool to convert to JSON, and edit a bit to get my own colors, descriptions etc. 

Only flaw with this, is that Google Maps sometimes routes you around traffic so the file you get by creating the route yourself, MAY be imprecise. But as long as I know this (I check as much as possible before making GPX-files), I'm good. 

![Skjermbilde 2021-06-19 kl  16 52 01](https://user-images.githubusercontent.com/4631971/122646360-06034100-d11f-11eb-87bf-577b04f5eba6.png)

I'll see if I take away the markers at each end. It may be too much with all the interstates in place. 

After all this is done, I'm going to try to use all my tracks from 11 trips in the US, and see if I can map them all on the same map. It's a mess as they are in Google maps, Swarm, old GPX-files, notebooks, EXIF data in images etc. 

But fun to use to learn.
